### ISSUE TITLE
Show Video Stats in an optional overlay window

### ISSUE THREAD

**Issue by Sascha-L** [Sascha-L/WPF-MediaKit#32]
With MediaPortal you can show the following stats:

https://forum.team-mediaportal.com/proxy.php?image=http%3A%2F%2Fs9.postimage.org%2F6ic9i3qz3%2FMP_stutter.png&hash=e1854d199d8c5ce62a57b94a626b8777

We should also add something like this, because it is very usefull to see if everything is working as expected :-)

**Comment by xmedeko** [comment]
@Sascha-L the link to the image is broken

**Comment by Sascha-L** [comment]
Fixed :D

**Comment by glenn-slayden** [comment]
One problem is that, in WPF, we don't have access to anywhere near that kind of detailed information about measured frame rate and the rest. Despite the fancy scheduling the WPF-MediaKit **EVRPresenter** tries to do (remember, our presenter is essentially identical to a Microsoft sample which was designed as a maximal example of various features), it's all wasted when we then use `D3DImage.SetBackBuffer` [D3DRenderer.cs#L370](/https://github.com/Sascha-L/WPF-MediaKit/blob/master/Source/DirectShow/Controls/D3DRenderer.cs#L370), which puts us at the mercy of WPF (re-)scheduling.

I've spent a great deal of time and research investigating how to overhaul this, with some tentatively positive results so far. For example:
 - Since the current approach to sample 'pooling' and 'scheduling' is overkill given WPF (as I noted above), these parts of our **EVRPresenter** can be drastically scaled back. In fact, in one experiment I've integrated an **EVRPresenter** written in C#, abandoning the C++ code altogether.
 - Synchronizing with the `Dispatcher` thread to set the back buffer is a huge issue also, so I think one of my most important improvements so far has been to use lock-free concurrency (`Interlocked` atomic operations) to transfer the `IDirect3DSurface9` across the native/managed boundary, which not only solves the tricky problems around having _code_ negotiate that transition, but also eliminates the need to synchronize with the UI thread once you're on the managed side.
 - And finally, I believe the only way to true get a grip on WPF VSync control is through some kind of runtime adaptive measurement/feedback loop using the Direct3D API `GetScanLine`,  as described [here](https://software.intel.com/en-us/articles/video-frame-display-synchronization). This part is still under investigation; I'm still not even sure it will be possible--even assuming you know exactly when the blanking interval starts--to **_reliably_** wake up WPF and such that you can set the back buffer in time.

But these are dramatic departures, and when my code is ready it may not resemble the WPF-MediaKit code base very much at all.

**Comment by xmedeko** [comment]
We do not need to display all of the informations as the mediaportal. We may start with the info we have: the renderer, the video and audio filter, the video size, etc. It would be handy to display some info when media are opening (e.g. "Media opening ... " is enough) and some error when media opening fails (just something like "Media opening error: {message}").

**Comment by xmedeko** [comment]
@glenn-slayden I do not understand this low level DirectX stuff much, but may this project https://github.com/Microsoft/WPFDXInterop help? (I just worry it's not maintained.)

**Comment by glenn-slayden** [comment]
Yes, I am aware of that project and it raises additional possibilities that I didn't happen to mention above, namely,  moving from Direct3D 9  to Direct3D 10 or 11. The approach detailed by the code on https://github.com/Microsoft/WPFDXInterop looks pretty straightforward and doesn't affect things about the more basic interop issues I'm looking at, so I figured it could wait as a lower priority.

**Comment by glenn-slayden** [comment]
Just to follow up my third bullet point above, VSync control is a lost cause on WPF--and in fact for DWM in general. Even if you could detect and wrangle the blanking interval, the user can trivially foil such delicate tuning efforts by adjusting the various VSync options in their (e.g.) Nvidia control panel. Furthermore, the whole effort might be vacuous if the user enables double- or triple-buffering in those aforementioned settings, and programatically detecting such scenarios in order to condition on them just complicates matters much further. Check [here](https://stackoverflow.com/a/42467599/147511) for more details.
